crux/src/Crux/Goal.hs

```diff
@@ -237,33 +239,66 @@ proveGoalsOffline adapter opts ctx explainFailure (Just gs0) = do
-                return (Prove (p, NotProved mempty Nothing))
-          case mres of
-            Just res -> return res
-            Nothing -> return (Prove (p, NotProved mempty Nothing))
 
+          let runOneSolver adapter = do
+                mres <- withTimeout $ WS.solver_adapter_check_sat adapter sym logData [assumptions, goal] $ \satRes ->
+                  case satRes of
+                    Unsat _ -> do
+                      -- NOTE: We don't have an easy way to get an unsat core here
+                      -- because we don't have a solver connection.
+                      let core = fmap Left (F.toList (mconcat assumptionsInScope))
+                      return (Proved core)
+                    Sat (evalFn, _) -> do
+                      let model = ctx ^. cruciblePersonality . personalityModel
+                      vals <- evalModel evalFn model
+                      explain <- explainFailure evalFn p
+                      return (NotProved explain (Just (ModelView vals)))
+                    Unknown -> do
+                      return (NotProved mempty Nothing)
+                case mres of
+                  Just res -> return res
+                  Nothing -> return (NotProved mempty Nothing)
+
+          res <- dispatchSolversOnGoalAsync adapters runOneSolver
+          end
+          case res of
+            Right details -> do
+              modifyIORef' goalNum (updateProcessedGoals p details)
+              case details of
+                NotProved _ (Just _) ->
+                  when (failfast && not (isResourceExhausted p)) $
+                    sayOK "Crux" "Counterexample found, skipping remaining goals"
+                _                  -> return ()
+              return (Prove (p, details))
+            Left es -> do
+              modifyIORef' goalNum (updateProcessedGoals p (NotProved mempty Nothing))
+              let allExceptions = unlines (displayException <$> es)
+              fail allExceptions
+                          
+dispatchSolversOnGoalAsync :: forall a s. [WS.SolverAdapter s]
+                      -> (WS.SolverAdapter s -> IO (ProofResult a))
+                      -> IO (Either [SomeException] (ProofResult a))
+dispatchSolversOnGoalAsync adapters withAdapter = do
+  asyncs <- forM adapters (async . withAdapter)
+  await asyncs []
+  where
+    await [] es = return $ Left es
+    await as es = do
+      (a, result) <- waitAnyCatch as
+      let as' = filter (/= a) as
+      case result of
+        Left  exc          -> await as' (exc : es)
+        Right r@(Proved _) -> do
+          mapM_ kill as'
+          return $ Right r
+        Right r@(NotProved _ (Just _)) -> do
+          mapM_ kill as'
+          return $ Right r
+        Right _ ->
+          await as' es
+
+    -- `cancel` from async blocks until the canceled thread has terminated.
+    kill a = throwTo (asyncThreadId a) AsyncCancelled
```

Comment:
Don't use an async exception here.  This generally bypasses exception handling code; but we want exception-handling to send signals to shutdown external solvers and have them shutdown.  `AsyncCancelled` will leave the solvers running and eating CPU time.  I tend to use `ExitSuccess`.